// vectra/lib/browser.js
"use strict";
/**
 * Browser entry point for Vectra.
 *
 * This module exports only browser-compatible components.
 * Node-specific components (LocalFileStorage, FileFetcher, WebFetcher) are excluded.
 *
 * For browser usage:
 * - Use `VirtualFileStorage` for in-memory storage (ephemeral)
 * - Use `IndexedDBStorage` for persistent browser storage
 * - Use `BrowserWebFetcher` for fetching web content
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.pathUtils = exports.BrowserWebFetcher = exports.IndexedDBStorage = exports.VirtualFileStorage = exports.FileStorageUtilities = exports.ItemSelector = exports.TransformersTokenizer = exports.TransformersEmbeddings = exports.OpenAIEmbeddings = exports.GPT3Tokenizer = exports.TextSplitter = exports.LocalDocumentResult = exports.LocalDocument = exports.LocalDocumentIndex = exports.LocalIndex = void 0;
// Core index classes
var LocalIndex_1 = require("./LocalIndex");
Object.defineProperty(exports, "LocalIndex", { enumerable: true, get: function () { return LocalIndex_1.LocalIndex; } });
var LocalDocumentIndex_1 = require("./LocalDocumentIndex");
Object.defineProperty(exports, "LocalDocumentIndex", { enumerable: true, get: function () { return LocalDocumentIndex_1.LocalDocumentIndex; } });
// Document classes
var LocalDocument_1 = require("./LocalDocument");
Object.defineProperty(exports, "LocalDocument", { enumerable: true, get: function () { return LocalDocument_1.LocalDocument; } });
var LocalDocumentResult_1 = require("./LocalDocumentResult");
Object.defineProperty(exports, "LocalDocumentResult", { enumerable: true, get: function () { return LocalDocumentResult_1.LocalDocumentResult; } });
// Text processing
var TextSplitter_1 = require("./TextSplitter");
Object.defineProperty(exports, "TextSplitter", { enumerable: true, get: function () { return TextSplitter_1.TextSplitter; } });
var GPT3Tokenizer_1 = require("./GPT3Tokenizer");
Object.defineProperty(exports, "GPT3Tokenizer", { enumerable: true, get: function () { return GPT3Tokenizer_1.GPT3Tokenizer; } });
// Embeddings
var OpenAIEmbeddings_1 = require("./OpenAIEmbeddings");
Object.defineProperty(exports, "OpenAIEmbeddings", { enumerable: true, get: function () { return OpenAIEmbeddings_1.OpenAIEmbeddings; } });
var TransformersEmbeddings_1 = require("./TransformersEmbeddings");
Object.defineProperty(exports, "TransformersEmbeddings", { enumerable: true, get: function () { return TransformersEmbeddings_1.TransformersEmbeddings; } });
var TransformersTokenizer_1 = require("./TransformersTokenizer");
Object.defineProperty(exports, "TransformersTokenizer", { enumerable: true, get: function () { return TransformersTokenizer_1.TransformersTokenizer; } });
// Query utilities
var ItemSelector_1 = require("./ItemSelector");
Object.defineProperty(exports, "ItemSelector", { enumerable: true, get: function () { return ItemSelector_1.ItemSelector; } });
var FileStorageUtilities_1 = require("./storage/FileStorageUtilities");
Object.defineProperty(exports, "FileStorageUtilities", { enumerable: true, get: function () { return FileStorageUtilities_1.FileStorageUtilities; } });
var VirtualFileStorage_1 = require("./storage/VirtualFileStorage");
Object.defineProperty(exports, "VirtualFileStorage", { enumerable: true, get: function () { return VirtualFileStorage_1.VirtualFileStorage; } });
var IndexedDBStorage_1 = require("./storage/IndexedDBStorage");
Object.defineProperty(exports, "IndexedDBStorage", { enumerable: true, get: function () { return IndexedDBStorage_1.IndexedDBStorage; } });
// Browser-compatible fetcher
var BrowserWebFetcher_1 = require("./BrowserWebFetcher");
Object.defineProperty(exports, "BrowserWebFetcher", { enumerable: true, get: function () { return BrowserWebFetcher_1.BrowserWebFetcher; } });
// Utilities
var pathUtils_1 = require("./utils/pathUtils");
Object.defineProperty(exports, "pathUtils", { enumerable: true, get: function () { return pathUtils_1.pathUtils; } });
