// vectra/lib/FolderWatcher.spec.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("assert");
const sinon = __importStar(require("sinon"));
const os = __importStar(require("os"));
const path = __importStar(require("path"));
const node_fs_1 = __importDefault(require("node:fs"));
const FolderWatcher_1 = require("./FolderWatcher");
const LocalDocumentIndex_1 = require("./LocalDocumentIndex");
const LocalFileStorage_1 = require("./storage/LocalFileStorage");
// Stub embeddings model that returns deterministic vectors
class StubEmbeddings {
    constructor() {
        this.maxTokens = 8000;
    }
    createEmbeddings(inputs) {
        return __awaiter(this, void 0, void 0, function* () {
            const texts = Array.isArray(inputs) ? inputs : [inputs];
            const output = texts.map(() => {
                const vec = new Array(384).fill(0);
                vec[0] = 1; // unit vector
                return vec;
            });
            return { status: 'success', output };
        });
    }
}
describe('FolderWatcher', () => {
    let tmpDir;
    let indexDir;
    let watchDir;
    let index;
    let sandbox;
    beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
        sandbox = sinon.createSandbox();
        tmpDir = yield node_fs_1.default.promises.mkdtemp(path.join(os.tmpdir(), 'vectra-watch-'));
        indexDir = path.join(tmpDir, 'index');
        watchDir = path.join(tmpDir, 'watch');
        yield node_fs_1.default.promises.mkdir(indexDir, { recursive: true });
        yield node_fs_1.default.promises.mkdir(watchDir, { recursive: true });
        index = new LocalDocumentIndex_1.LocalDocumentIndex({
            folderPath: indexDir,
            embeddings: new StubEmbeddings(),
            storage: new LocalFileStorage_1.LocalFileStorage(),
        });
        yield index.createIndex({ version: 1, deleteIfExists: true });
    }));
    afterEach(() => __awaiter(void 0, void 0, void 0, function* () {
        sandbox.restore();
        yield node_fs_1.default.promises.rm(tmpDir, { recursive: true, force: true });
    }));
    it('should perform initial sync of existing files', () => __awaiter(void 0, void 0, void 0, function* () {
        // Create files before starting watcher
        yield node_fs_1.default.promises.writeFile(path.join(watchDir, 'file1.txt'), 'hello world');
        yield node_fs_1.default.promises.writeFile(path.join(watchDir, 'file2.txt'), 'goodbye world');
        const watcher = new FolderWatcher_1.FolderWatcher({ index, paths: [watchDir] });
        const synced = [];
        watcher.on('sync', (uri) => synced.push(uri));
        yield watcher.start();
        try {
            assert_1.strict.equal(watcher.trackedFileCount, 2);
            assert_1.strict.equal(synced.length, 2);
            // Verify documents exist in index
            const id1 = yield index.getDocumentId(path.join(watchDir, 'file1.txt'));
            const id2 = yield index.getDocumentId(path.join(watchDir, 'file2.txt'));
            assert_1.strict.ok(id1, 'file1.txt should be indexed');
            assert_1.strict.ok(id2, 'file2.txt should be indexed');
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should emit ready event after initial sync', () => __awaiter(void 0, void 0, void 0, function* () {
        const watcher = new FolderWatcher_1.FolderWatcher({ index, paths: [watchDir] });
        let ready = false;
        watcher.on('ready', () => { ready = true; });
        yield watcher.start();
        try {
            assert_1.strict.equal(ready, true);
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should filter by extensions', () => __awaiter(void 0, void 0, void 0, function* () {
        yield node_fs_1.default.promises.writeFile(path.join(watchDir, 'include.txt'), 'included');
        yield node_fs_1.default.promises.writeFile(path.join(watchDir, 'exclude.js'), 'excluded');
        yield node_fs_1.default.promises.writeFile(path.join(watchDir, 'include.md'), 'also included');
        const watcher = new FolderWatcher_1.FolderWatcher({
            index,
            paths: [watchDir],
            extensions: ['.txt', '.md']
        });
        yield watcher.start();
        try {
            assert_1.strict.equal(watcher.trackedFileCount, 2);
            const idTxt = yield index.getDocumentId(path.join(watchDir, 'include.txt'));
            const idMd = yield index.getDocumentId(path.join(watchDir, 'include.md'));
            const idJs = yield index.getDocumentId(path.join(watchDir, 'exclude.js'));
            assert_1.strict.ok(idTxt, 'include.txt should be indexed');
            assert_1.strict.ok(idMd, 'include.md should be indexed');
            assert_1.strict.equal(idJs, undefined, 'exclude.js should not be indexed');
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should handle extensions without leading dot', () => __awaiter(void 0, void 0, void 0, function* () {
        yield node_fs_1.default.promises.writeFile(path.join(watchDir, 'test.txt'), 'hello');
        const watcher = new FolderWatcher_1.FolderWatcher({
            index,
            paths: [watchDir],
            extensions: ['txt']
        });
        yield watcher.start();
        try {
            assert_1.strict.equal(watcher.trackedFileCount, 1);
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should recurse into subdirectories', () => __awaiter(void 0, void 0, void 0, function* () {
        const subDir = path.join(watchDir, 'sub');
        yield node_fs_1.default.promises.mkdir(subDir, { recursive: true });
        yield node_fs_1.default.promises.writeFile(path.join(watchDir, 'root.txt'), 'root');
        yield node_fs_1.default.promises.writeFile(path.join(subDir, 'nested.txt'), 'nested');
        const watcher = new FolderWatcher_1.FolderWatcher({ index, paths: [watchDir] });
        yield watcher.start();
        try {
            assert_1.strict.equal(watcher.trackedFileCount, 2);
            const idRoot = yield index.getDocumentId(path.join(watchDir, 'root.txt'));
            const idNested = yield index.getDocumentId(path.join(subDir, 'nested.txt'));
            assert_1.strict.ok(idRoot, 'root.txt should be indexed');
            assert_1.strict.ok(idNested, 'nested.txt should be indexed');
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should watch individual files', () => __awaiter(void 0, void 0, void 0, function* () {
        const singleFile = path.join(tmpDir, 'single.txt');
        yield node_fs_1.default.promises.writeFile(singleFile, 'single file');
        const watcher = new FolderWatcher_1.FolderWatcher({ index, paths: [singleFile] });
        yield watcher.start();
        try {
            assert_1.strict.equal(watcher.trackedFileCount, 1);
            const id = yield index.getDocumentId(singleFile);
            assert_1.strict.ok(id, 'single.txt should be indexed');
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should watch multiple paths', () => __awaiter(void 0, void 0, void 0, function* () {
        const dir2 = path.join(tmpDir, 'watch2');
        yield node_fs_1.default.promises.mkdir(dir2, { recursive: true });
        yield node_fs_1.default.promises.writeFile(path.join(watchDir, 'a.txt'), 'a');
        yield node_fs_1.default.promises.writeFile(path.join(dir2, 'b.txt'), 'b');
        const watcher = new FolderWatcher_1.FolderWatcher({ index, paths: [watchDir, dir2] });
        yield watcher.start();
        try {
            assert_1.strict.equal(watcher.trackedFileCount, 2);
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should handle sync() detecting deleted files', () => __awaiter(void 0, void 0, void 0, function* () {
        yield node_fs_1.default.promises.writeFile(path.join(watchDir, 'ephemeral.txt'), 'temporary');
        const watcher = new FolderWatcher_1.FolderWatcher({ index, paths: [watchDir] });
        yield watcher.start();
        assert_1.strict.equal(watcher.trackedFileCount, 1);
        // Delete the file
        yield node_fs_1.default.promises.unlink(path.join(watchDir, 'ephemeral.txt'));
        // Manual sync should detect deletion
        const synced = [];
        watcher.on('sync', (uri, action) => synced.push({ uri, action }));
        yield watcher.sync();
        try {
            assert_1.strict.equal(watcher.trackedFileCount, 0);
            const deleted = synced.find(s => s.action === 'deleted');
            assert_1.strict.ok(deleted, 'should have emitted a delete event');
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should handle sync() detecting updated files', () => __awaiter(void 0, void 0, void 0, function* () {
        const filePath = path.join(watchDir, 'mutable.txt');
        yield node_fs_1.default.promises.writeFile(filePath, 'version 1');
        const watcher = new FolderWatcher_1.FolderWatcher({ index, paths: [watchDir] });
        yield watcher.start();
        assert_1.strict.equal(watcher.trackedFileCount, 1);
        // Update the file (ensure mtime changes)
        yield new Promise(r => setTimeout(r, 50));
        yield node_fs_1.default.promises.writeFile(filePath, 'version 2');
        // Manual sync should detect update
        const synced = [];
        watcher.on('sync', (uri, action) => synced.push({ uri, action }));
        yield watcher.sync();
        try {
            assert_1.strict.equal(watcher.trackedFileCount, 1);
            const updated = synced.find(s => s.action === 'updated');
            assert_1.strict.ok(updated, 'should have emitted an update event');
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should not throw if path does not exist', () => __awaiter(void 0, void 0, void 0, function* () {
        const watcher = new FolderWatcher_1.FolderWatcher({
            index,
            paths: [path.join(tmpDir, 'nonexistent')]
        });
        yield watcher.start();
        try {
            assert_1.strict.equal(watcher.trackedFileCount, 0);
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should throw if started twice', () => __awaiter(void 0, void 0, void 0, function* () {
        const watcher = new FolderWatcher_1.FolderWatcher({ index, paths: [watchDir] });
        yield watcher.start();
        try {
            yield assert_1.strict.rejects(() => watcher.start(), /already running/);
        }
        finally {
            yield watcher.stop();
        }
    }));
    it('should report isRunning correctly', () => __awaiter(void 0, void 0, void 0, function* () {
        const watcher = new FolderWatcher_1.FolderWatcher({ index, paths: [watchDir] });
        assert_1.strict.equal(watcher.isRunning, false);
        yield watcher.start();
        assert_1.strict.equal(watcher.isRunning, true);
        yield watcher.stop();
        assert_1.strict.equal(watcher.isRunning, false);
    }));
    it('should emit error events for sync failures', () => __awaiter(void 0, void 0, void 0, function* () {
        yield node_fs_1.default.promises.writeFile(path.join(watchDir, 'bad.txt'), 'content');
        // Create watcher with no embeddings to force error
        const badIndex = new LocalDocumentIndex_1.LocalDocumentIndex({
            folderPath: indexDir,
            // no embeddings — will throw on upsertDocument
            storage: new LocalFileStorage_1.LocalFileStorage(),
        });
        const watcher = new FolderWatcher_1.FolderWatcher({ index: badIndex, paths: [watchDir] });
        const errors = [];
        watcher.on('error', (err, uri) => errors.push({ err, uri }));
        yield watcher.start();
        try {
            assert_1.strict.equal(errors.length, 1);
            assert_1.strict.ok(errors[0].err.message.includes('Embeddings model not configured'));
        }
        finally {
            yield watcher.stop();
        }
    }));
});
